---
title: "Filter Target Loci and Generate AGG PAMs, +5 G to T edits"
author: "Katya Williams"
date: "`r Sys.Date()`"
output: html_document
---

# 1. Setup

## 1.1 Load necessary libraries

```{r setup}
# Load any necessary libraries
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(seqinr)
```

## 1.2 Read in top target site candidate dataset

*This script creates +5 G to T edits, so it should only be used with `MMR-responsive_+5GtoT_target_candidates.csv`*

```{r read-plus5-G-to-T}
top <- read_csv('MMR-responsive_+5GtoT_target_candidates.csv')
```

## 1.3 Read in any auxiliary datasets

Read in the joined version of lDS004 data

```{r read-aux}
diff <- read.csv('lDS004_joined_mean.csv')
```

# 2. Filter candidate target sites

## 2.1 Generate truncated target sequences

Create truncated target sequences along with a non-editing RT (modified from Ann's RTs)

```{r oligos}
# initialize empty vectors
RT_revert <- vector(length=nrow(top))
PAM <- vector(length=nrow(top))
Target <- vector(length=nrow(top))
coding_seq <- vector(length=nrow(top))

# change 5th nt in RT_revcomp to value of Edited_nt
for(i in 1:nrow(top)) {
    # get reverse complement of RT
    RT_revcomp <- toupper(paste(rev(comp(unlist(strsplit(top$RT_template[i], '')))), collapse=''))
    # extract PAM sequence from reverse complement RT
    ## PAM encoded as reverse complement at end of RT sequence (bases 4-6)
    ## need to change 5th base in reverse complement back to original
    split <- unlist(strsplit(RT_revcomp, ""))
    split[5] <- top$Reference_nt[i]
    RT_revert_rc <- paste(split, collapse='')
    # generate unedited RT template
    RT_revert[i] <- toupper(paste(rev(comp(unlist(strsplit(RT_revert_rc, '')))), collapse=''))
    # generate PAM alone
    PAM[i] <- toupper(substr(RT_revert_rc, 4, 6))
    
    # find the index of the nt immediately after the end of the PAM sequence in the wide target
    after_start <- unlist(gregexpr(top$Protospacer[i], top$Wide_target_sequence[i], fixed=TRUE)) + 23
    # find the length of the end of the RT template in the wide target
    after_endlen <- top$RT_length[i] - 7
    # substring the wide target nt between the PAM and end of the RT -- these are necessary for full RT binding
    after_PAM <- substr(top$Wide_target_sequence[i], after_start, after_start + after_endlen)
    # store the start of the coding sequence (concatenate edited PAM and after_PAM)
    coding_seq[i] <- paste(toupper(c('ATG', after_PAM)), collapse='')
    # generate minimal target sequence (protospacer + PAM)
    Target[i] <- paste(toupper(c(top$Protospacer[i], PAM[i], after_PAM)), collapse='')
}

# add columns to dataset
top <- top %>% mutate(RT_revert = RT_revert,
                      PAM = PAM,
                      Target = Target,
                      Start_of_Coding_Seq = coding_seq)
```

## 2.2 Truncated target filtering

### 2.2.1 Remove targets already containing start codons

Otherwise, reporter will be active regardless of editing

**This code only identifies start codons in the forward direction, where the PAM is on the coding (sense) strand.** Reverse-oriented target sequences are incompatible with this method since the PAM would be on the nonsense strand and thus would not function as a start codon, but a reverse-facing target would allow strand invasion away from the reporter sequence and introduce fewer indels into the coding sequence.

```{r valid-targets}
## forward-oriented edits: look for ATG in target sequence

# check for start codons in the target seq (protospacer + PAM)
top <- top %>% mutate(Target_hasATG = grepl('ATG', top$Target, fixed = TRUE))

# generate table of only start codon-less targets
peg_library_start <- top %>% 
    filter(Target_hasATG == FALSE)

# using a reverse-oriented target sequences is incompatible with this method since the PAM needs to be on the sense strand, but a reverse-facing target would allow strand invasion away from the reporter sequence and introduce fewer indels into the coding sequence

length(peg_library_start)

# how many targets already contain an ATG?
targets_miss <- setdiff(top$Target_Locus, peg_library_start$Target_Locus)
length(targets_miss)
```

### 2.2.2 Remove any edits where the coding sequence contains a stop codon after the PAM from the data

Otherwise, the reporter will be inactive regardless of editing: the reporter cannot be translated, even if a start codon is successfully introduced

```{r remove-stop}
codon_list <- vector(length=nrow(peg_library_start))

# split sequence into codons
for (i in 1:nrow(peg_library_start)) {
    # split the seq into a vector of individual characters
    seq <- unlist(strsplit(peg_library_start$Start_of_Coding_Seq[i], ''))
    codons <- splitseq(seq, frame=0, word=3)
    codon_list[i] <- toString(codons)
}

peg_library_start <- peg_library_start %>% mutate(Starting_Codons = codon_list)

# search for stop codons in Start_of_Coding_Seq
peg_library_start <- peg_library_start %>% mutate(Coding_hasUAA = grepl('TAA', peg_library_start$Starting_Codons, fixed = TRUE),
                                      Coding_hasUAG = grepl('TAG', peg_library_start$Starting_Codons, fixed = TRUE),
                                      Coding_hasUGA = grepl('TGA', peg_library_start$Starting_Codons, fixed = TRUE))

# storage vector for for loop
anystops <- c()

# for each row...
for (i in 1:nrow(peg_library_start)) {
    # are any of the stop codons present?
    stops <- any(c(peg_library_start$Coding_hasUAA[i] == TRUE, 
                   peg_library_start$Coding_hasUAG[i] == TRUE,
                   peg_library_start$Coding_hasUGA[i] == TRUE))
    anystops <- c(anystops, stops)
}

peg_library_start <- peg_library_start %>% mutate(Coding_hasStop = anystops)

# remove any data where Coding_hasStop == TRUE
peg_library_stops <- peg_library_start %>% filter(Coding_hasStop == FALSE)
```

# 3. Create AGG PAMs and +5 G to T RTs

# 3.1 Change the 21st nt in all target sequences to A, and make +5GtoT edits

```{r new-target}
# make storage vectors
new_target <- vector(length=nrow(peg_library_stops))
new_target_wide <- vector(length=nrow(peg_library_stops))
new_rt <- vector(length=nrow(peg_library_stops))
edit_pos <- vector(length=nrow(peg_library_stops))
ref_nt <- vector(length=nrow(peg_library_stops))
edited_nt <- vector(length=nrow(peg_library_stops))
edited_target <- vector(length=nrow(peg_library_stops))
edited_target_wide <- vector(length=nrow(peg_library_stops))
peg_id <- vector(length=nrow(peg_library_stops))
is_A <- vector(length=nrow(peg_library_stops))

# for each potential edit...
for (i in 1:nrow(peg_library_stops)) {
    # what's the target locus?
    target_locus <- peg_library_stops$Target_Locus[i]
    # what's the PBS length?
    pbslen <- peg_library_stops$PBS_length[i]
    # what's the RT length?
    rtlen <- peg_library_stops$RT_length[i]
    
    # change the 21st nt to A in the truncated target
    split <- unlist(strsplit(peg_library_stops$Target[i], ""))
    isA <- split[21] == 'A' # check to see if the nt is already A
    split[21] <- 'A'
    new_target[i] <- paste(toupper(split), collapse='')
    
    # change the 21st nt to A in the wide target
    split_wide <- unlist(strsplit(peg_library_stops$Wide_target_sequence[i], ""))
    split_wide[25] <- 'A'
    new_target_wide[i] <- paste(toupper(split_wide), collapse='')
    
    # and then change the 22nd nt to T to generaate edited targets
    split[22] <- 'T'
    edited_target[i] <- paste(toupper(split), collapse='')
    split_wide[26] <- 'T'
    edited_target_wide[i] <- paste(toupper(split_wide), collapse='')
    
    # get the reverse complement of the RT template (edit is always a set distance from the end of the RT, so temporarily make the end the start for computational purposes)
    split_rtrc <- rev(comp(unlist(strsplit(peg_library_stops$RT_template[i], ''))))
    # change the 4th nt to A and the 5th nt to T
    split_rtrc[4] <- 'A'
    split_rtrc[5] <- 'T'
    new_rt[i] <- paste(toupper(rev(comp(split_rtrc))), collapse='')
    
    # generate peg ID
    peg_id[i] <- paste0(target_locus, '_+5GtoT_', pbslen, '_', rtlen)
    
    # fill in the other storage vectors
    edit_pos[i] <- 5
    ref_nt[i] <- 'G'
    edited_nt[i] <- 'T'
    is_A[i] <- isA
}
```

# 3.2 Build a table of detailed information on each edit

```{r new-table}
# build a new table from scratch (so column order makes sense)
peg_library <- tibble(pegRNA_ID = peg_id, Gene = peg_library_stops$Gene,
                      Target_Locus = peg_library_stops$Target_Locus,
                      Guide = peg_library_stops$Guide,
                      Protospacer = peg_library_stops$Protospacer,
                      PBS = peg_library_stops$PBS, Anns_RT = peg_library_stops$RT_template,
                      RT_template = new_rt, Barcode = peg_library_stops$Barcode,
                      Original_Target = peg_library_stops$Target, Target_4A = new_target,
                      Edited_Target = edited_target,
                      Original_Wide_Target = peg_library_stops$Wide_target_sequence,
                      Wide_Target_4A = new_target_wide,
                      Edited_Wide_Target = edited_target_wide,
                      Target_had_4A = is_A, Edit_position = edit_pos, Reference_nt = ref_nt,
                      Edited_nt = edited_nt, PBS_length = peg_library_stops$PBS_length,
                      RT_length = peg_library_stops$RT_length,
                      GC_percent_PBS = peg_library_stops$GC_percent_PBS,
                      MLH1_Avg_Freq_TotalEdited_d7 = peg_library_stops$MLH1_Avg_Freq_TotalEdited_d7, 
                      PEmax_Avg_Freq_TotalEdited_d7 = peg_library_stops$PEmax_Avg_Freq_TotalEdited_d7, 
                      MLH1_Avg_Freq_EditOnly_d7 = peg_library_stops$MLH1_Avg_Freq_EditOnly_d7, 
                      PEmax_Avg_Freq_EditOnly_d7 = peg_library_stops$PEmax_Avg_Freq_EditOnly_d7, 
                      MLH1_Avg_Freq_TotalErrors = peg_library_stops$MLH1_Avg_Freq_TotalErrors, 
                      PEmax_Avg_Freq_TotalErrors = peg_library_stops$PEmax_Avg_Freq_TotalErrors, 
                      MLH1_Avg_TotalReads = peg_library_stops$MLH1_Avg_TotalReads,
                      PEmax_Avg_TotalReads = peg_library_stops$PEmax_Avg_TotalReads,
                      mean_diff_avg_freq = peg_library_stops$mean_diff_avg_freq,
                      Target_hasATG = peg_library_stops$Target_hasATG,
                      Coding_hasUAA = peg_library_stops$Coding_hasUAA,
                      Coding_hasUAG = peg_library_stops$Coding_hasUAG,
                      Coding_hasUGA = peg_library_stops$Coding_hasUGA,
                      Coding_hasStop = peg_library_stops$Coding_hasStop,
                      Start_of_Coding_Seq = peg_library_stops$Start_of_Coding_Seq)
```

# 4. Adjust truncated targets for in-frame ATG

## 4.1 Make a list of codons in preferred order

Ranking hierarchy:
1) nonpolar over polar uncharged over aromatic over polar charged sidechain
2) small sidechain over large sidechain
3) codon likelihood

Stop and start codons are removed from this list to a) ensure successful editing enables translation of the reporter sequence and b) ensure *only* succcessful editing enables translation of the reporter sequence

```{r make-codon-table}
codon_rank <- c('GGC', 'GGA', 'GGG', 'GGT', 'GCC', 'GCT', 'GCA', 'GCG', 'GTG', 'GTC', 'GTT', 'GTA', 'CTG', 'CTC', 'CTT', 'TTG', 'TTA', 'CTA', 'ATC', 'ATT', 'ATA', 'AGC', 'TCC', 'TCT', 'AGT', 'TCA', 'TCG', 'ACC', 'ACA', 'ACT', 'ACG', 'TGC', 'TGT', 'AAC', 'AAT', 'CAG', 'CAA', 'CCC', 'CCT', 'CCA', 'CCG', 'TTC', 'TTT', 'TAC', 'TAT', 'TGG', 'GAC', 'GAT', 'GAG', 'GAA', 'AAG', 'AAA', 'AGA', 'AGG', 'CGG', 'CGC', 'CGA', 'CGT', 'CAC', 'CAT')
amino_rank <- c('Gly', 'Gly', 'Gly', 'Gly', 'Ala', 'Ala', 'Ala', 'Ala', 'Val', 'Val', 'Val', 'Val', 'Leu', 'Leu', 'Leu', 'Leu', 'Leu', 'Leu', 'Ile', 'Ile', 'Ile', 'Ser', 'Ser', 'Ser', 'Ser', 'Ser', 'Ser', 'Thr', 'Thr', 'Thr', 'Thr', 'Cys', 'Cys', 'Asn', 'Asn', 'Gln', 'Gln', 'Pro', 'Pro', 'Pro', 'Pro', 'Phe', 'Phe', 'Tyr', 'Tyr', 'Trp', 'Asp', 'Asp', 'Glu', 'Glu', 'Lys', 'Lys', 'Arg', 'Arg', 'Arg', 'Arg', 'Arg', 'Arg', 'His', 'His')

codon_ranking <- tibble(Codon = codon_rank, Amino_Acid = amino_rank)
write.csv(codon_ranking, 'Ranked_Codon_List.csv', row.names = TRUE)
```

## 4.2 Check to make sure the edit is in-frame: if not, find the number of nucleotides to append

```{r inframe}
# create storage vectors
n_toappend <- c() # the number of nt to append to make the ATG in-frame

# for each edited minimal target sequence...
for (i in 1:nrow(peg_library)) {
    # split the target into a vector of individual characters
    split_mintar <- unlist(strsplit(peg_library$Edited_Target[i], ''))

    # find length of target
    min_len <- length(split_mintar)
    
    # find start of ATG
    ATG_start <- unlist(gregexpr('ATG', peg_library$Edited_Target[i], fixed=TRUE))

    # subtract end of ATG index from the target length
    dist_nluc <- min_len - (ATG_start + 2)
        
    # find dist_nluc modulo 3 to determine if ATG is in-frame
    leftover <- dist_nluc %% 3
    
    # subtract the leftover from 3 to get the number of nt to append
    toappend <- 3 - leftover
    
    # if toappend is 3, change its value to 0: the edit is in frame
    if (toappend == 3) toappend <- 0
        
    # add to the storage vectors
    n_toappend <- c(n_toappend, toappend)
}

# add the storage vectors to the library data
peg_library <- peg_library %>% mutate(Number_to_Append = n_toappend)
```

## 4.3 Append nucleotides to truncated target sequences

If nt need to be appended,
1) find the current nt
2) determine the best nt to append based on outcomes of current nts using ranked codon list

```{r nt-toappend}
# create storage vectors
nt_toappend <- vector(length=nrow(peg_library))
inframe_seq <- vector(length=nrow(peg_library))
inframe_target <- vector(length=nrow(peg_library))
inframe_edited_target <- vector(length=nrow(peg_library))

# for each edit...
for (i in 1:nrow(peg_library)) {
    # if Number_to_Append == 0...
    if (peg_library$Number_to_Append[i] == 0) {
        # store nothing in the appending vector and duplicate the unappended value
        nt_toappend[i] <- '-'
        inframe_seq[i] <- peg_library$Start_of_Coding_Seq[i]
        inframe_target[i] <- peg_library$Target_4A[i]
        inframe_edited_target[i] <- peg_library$Edited_Target[i]
    } else {
        # parse the last Number_to_Append nt
        toappend <- peg_library$Number_to_Append[i]
        
        # extract the last 3 - Number_to_Append nt from Start_of_Coding_Seq
        len <- length(unlist(strsplit(peg_library$Start_of_Coding_Seq[i], '')))
        endseq <- toupper(substr(peg_library$Start_of_Coding_Seq[i], len-(3-toappend)+1, len))
        
        # make a string of '.' to append to fill out the codon
        if (toappend == 1) {
            dots <- '.'
        } else {
            dots <- '..'
        }
        
        # append dots to make a full codon search term
        search <- paste0(endseq, dots, collapse='')
        
        # match the nt to the first matching possible codon in the ranked codon list
        df <- codon_ranking[grep(search, codon_ranking$Codon),]
        match <- df$Codon[1]
        
        # parse the last 3 - Number_to_Append of this codon and store
        if (toappend == 1) {
            addnt <- toupper(substr(match, 3, 3))
        } else {
            addnt <- toupper(substr(match, 2, 3))
        }
        nt_toappend[i] <- addnt
        # concatenate these nt to the end of the sequences and store
        inframe_seq[i] <- paste0(peg_library$Start_of_Coding_Seq[i], addnt, collapse='')

        # editing is being done to the sense strand -- just add nt's
        inframe_target[i] <- paste0(peg_library$Target_4A[i], addnt, collapse='')
        inframe_edited_target[i]  <- paste0(peg_library$Edited_Target[i], addnt, collapse='')
    }
}

# add columns to the dataset
peg_library <- peg_library %>% mutate(Nt_to_Append = nt_toappend,
                                      In_Frame_Coding_Seq = inframe_seq,
                                      In_Frame_Target_4A = inframe_target,
                                      In_Frame_Edited_Target = inframe_edited_target)
```

## 4.4 Generate amino acid sequence 

```{r translate}
# storage vector for translated sequences
aa_seqs <- vector(length=nrow(peg_library))

# for each row in peg_library...
for (i in 1:nrow(peg_library)) {
    # store the in-frame start of the coding sequence as a DNAstring object
    dnaseq <- Biostrings::DNAString(peg_library$In_Frame_Coding_Seq[i])
    
    # translate the sequence
    aaseq <- toString(Biostrings::translate(dnaseq))
    
    # store the translated sequence
    aa_seqs[i] <- aaseq
}

# add amino acid sequence column to peg_library
peg_library <- peg_library %>% mutate(AA_Sequence = aa_seqs)
```

```{r add-edits}
peg_library$Edit_nts <- paste(peg_library$Reference_nt, '>', peg_library$Edited_nt)
```

# 5. Add controls

Manually add controls to the data

```{r controls}
# write out column values for control rows
nontarget1 <- c(pegRNA_ID =	'non-targeting_647_NA.to._11_10', Gene = 'non-targeting', 
                Target_Locus = 'non-targeting', Guide = 'GGCCGGCTTGTGACAGTGAA', 
                Protospacer = 'AGCCGGCTTGTGACAGTGAA', 
                PBS = 'ACTGTCACAAG', Anns_RT = 'AGCATTATTC', RT_template = 'AGCATTATTC',
                Barcode = 'AGTGCCTTGCTGCCTCC',
                Original_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Target_4A = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG', 
                Edited_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Original_Wide_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Wide_Target_4A = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Edited_Wide_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Target_had_4A = NA, Edit_position = NA, Reference_nt = NA, Edited_nt = NA,
                PBS_length = 11, RT_length = 10, GC_percent_PBS = 0.45454545, 
                MLH1_Avg_Freq_TotalEdited_d7 = 0, PEmax_Avg_Freq_TotalEdited_d7 = 0,
                MLH1_Avg_Freq_EditOnly_d7 = 0, PEmax_Avg_Freq_EditOnly_d7 = 0,
                MLH1_Avg_Freq_TotalErrors = 0.034278191, 
                PEmax_Avg_Freq_TotalErrors = 0.028298577, 
                MLH1_Avg_TotalReads = 18063.5, PEmax_Avg_TotalReads = 20193,
                mean_diff_avg_freq = 0, Target_hasATG = TRUE, Coding_hasUAA = NA,
                Coding_hasUAG = NA, Coding_hasUGA = NA, Coding_hasStop = NA,
                Start_of_Coding_Seq = NA, Number_to_Append = NA, Nt_to_Append = NA,
                In_Frame_Coding_Seq = NA, 
                In_Frame_Target_4A = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                In_Frame_Edited_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                AA_Sequence = NA, Edit_nts = 'negative_control')

nontarget2 <- c(pegRNA_ID =	'non-targeting_650_NA.to._13_15', Gene = 'non-targeting', 
                Target_Locus = 'non-targeting', Guide = 'GACTGCAGTATTCGTGGCCT', 
                Protospacer = 'CACTGCAGTATTCGTGGCCT', 
                PBS = 'CCACGAATACTGC', Anns_RT = 'AATGTGACGGGGAGG', 
                RT_template = 'AATGTGACGGGGAGG', Barcode = 'GGCGCGTCCCGCGAAGG',
                Original_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Target_4A = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG', 
                Edited_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Wide_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Wide_Target_4A = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Edited_Wide_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                Target_had_4A = NA, Edit_position = NA, Reference_nt = NA, Edited_nt = NA,
                PBS_length = 13, RT_length = 15, GC_percent_PBS = 0.53846154, 
                MLH1_Avg_Freq_TotalEdited_d7 = 0, PEmax_Avg_Freq_TotalEdited_d7 = 0,
                MLH1_Avg_Freq_EditOnly_d7 = 0, PEmax_Avg_Freq_EditOnly_d7 = 0,
                MLH1_Avg_Freq_TotalErrors = 0.019864869, 
                PEmax_Avg_Freq_TotalErrors = 0.019240683, 
                MLH1_Avg_TotalReads = 11192.5, PEmax_Avg_TotalReads = 12489.5,
                mean_diff_avg_freq = 0, Target_hasATG = TRUE, Coding_hasUAA = NA,
                Coding_hasUAG = NA, Coding_hasUGA = NA, Coding_hasStop = NA,
                Start_of_Coding_Seq = NA, Number_to_Append = NA, Nt_to_Append = NA,
                In_Frame_Coding_Seq = NA, 
                In_Frame_Target_4A = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                In_Frame_Edited_Target = 'GGCGAGTTACTCTCGTAATAGGAATCCTCTTGCATGTACATGAAGTG',
                AA_Sequence = NA, Edit_nts = 'negative_control')

HEK3_noedit <- c(pegRNA_ID = 'HEK3_645_--to-_13_10', Gene = 'HEK3', 
                 Target_Locus = 'HEK3_645', Guide = 'GGCCCAGACTGAGCACGTGA', 
                 Protospacer = 'GGCCCAGACTGAGCACGTGA', 
                 PBS = 'CGTGCTCAGTCTG', Anns_RT = 'TCTGCCATCA', RT_template = 'TCTGCCATCA',
                 Barcode = 'AACGCCCGTAATAGGCG',
                 Original_Target = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC',
                 Target_4A = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC', 
                 Edited_Target = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC',
                 Original_Wide_Target = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC',
                 Wide_Target_4A = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC', 
                 Edited_Wide_Target = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC',
                 Target_had_4A = NA, Edit_position = NA, Reference_nt = NA, Edited_nt = NA,
                 PBS_length = 13, RT_length = 10, GC_percent_PBS = 0.61538462, 
                 MLH1_Avg_Freq_TotalEdited_d7 = 0, PEmax_Avg_Freq_TotalEdited_d7 = 0,
                 MLH1_Avg_Freq_EditOnly_d7 = 0, PEmax_Avg_Freq_EditOnly_d7 = 0,
                 MLH1_Avg_Freq_TotalErrors = 0.015458493, 
                 PEmax_Avg_Freq_TotalErrors = 0.013183324, 
                 MLH1_Avg_TotalReads = 20869, PEmax_Avg_TotalReads = 23135,
                 mean_diff_avg_freq = 0, Target_hasATG = TRUE, Coding_hasUAA = NA,
                 Coding_hasUAG = NA, Coding_hasUGA = NA, Coding_hasStop = NA,
                 Start_of_Coding_Seq = NA, Number_to_Append = NA, Nt_to_Append = NA,
                 In_Frame_Coding_Seq = NA, 
                 In_Frame_Target_4A = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC',
                 In_Frame_Edited_Target = 'TTGGGGCCCAGACTGAGCACGTGATGGCAGAGGAAAGGAAGCCCTGC',
                 AA_Sequence = NA, Edit_nts = 'negative_control')

VEGFA <- c(pegRNA_ID = 'VEGFA_641_5GtoT_13_22', Gene = 'VEGFA', 
                 Target_Locus = 'VEGFA_641', Guide = 'GATGTCTGCAGGCCAGATGA', 
                 Protospacer = 'GATGTCTGCAGGCCAGATGA', 
                 PBS = 'TCTGGCCTGCAGA', Anns_RT = 'AATGTGCCATCTGGAGCACTCA', 
                 RT_template = 'AATGTGCCATCTGGAGCATTCA',
                 Barcode = 'GCTAATATGGGCGCCGC',
                 Original_Target = 'CTTTGATGTCTGCAGGCCAGATGAGGGCTCCAGATGGCACATTGTCA',
                 Target_4A =       'CTTTGATGTCTGCAGGCCAGATGAAGGCTCCAGATGGCACATTGTCA', 
                 Edited_Target =   'CTTTGATGTCTGCAGGCCAGATGAATGCTCCAGATGGCACATTGTCA',
                 Original_Wide_Target = 'CTTTGATGTCTGCAGGCCAGATGAGGGCTCCAGATGGCACATTGTCA',
                 Wide_Target_4A = 'CTTTGATGTCTGCAGGCCAGATGAAGGCTCCAGATGGCACATTGTCA', 
                 Edited_Wide_Target = 'CTTTGATGTCTGCAGGCCAGATGAATGCTCCAGATGGCACATTGTCA',
                 Target_had_4A = NA, Edit_position = 5, Reference_nt = 'G', Edited_nt = 'T',
                 PBS_length = 13, RT_length = 22, GC_percent_PBS = 50, 
                 MLH1_Avg_Freq_TotalEdited_d7 = 0.86388162, 
                 PEmax_Avg_Freq_TotalEdited_d7 = 0.267906784,
                 MLH1_Avg_Freq_EditOnly_d7 = 0.84749969, 
                 PEmax_Avg_Freq_EditOnly_d7 = 0.267906784,
                 MLH1_Avg_Freq_TotalErrors = 0.01638193, 
                 PEmax_Avg_Freq_TotalErrors = 0.009087295, 
                 MLH1_Avg_TotalReads = 5213.5, PEmax_Avg_TotalReads = 5494.5,
                 mean_diff_avg_freq = 0.579592906, Target_hasATG = TRUE, Coding_hasUAA = NA,
                 Coding_hasUAG = NA, Coding_hasUGA = NA, Coding_hasStop = NA,
                 Start_of_Coding_Seq = NA, Number_to_Append = NA, Nt_to_Append = NA,
                 In_Frame_Coding_Seq = NA, 
                 In_Frame_Target_4A =     'CTTTGATGTCTGCAGGCCAGATGAAGGCTCCAGATGGCACATTGTCA',
                 In_Frame_Edited_Target = 'CTTTGATGTCTGCAGGCCAGATGAATGCTCCAGATGGCACATTGTCA',
                 AA_Sequence = NA, Edit_nts = 'positive_control')

RUNX1 <- c(pegRNA_ID = 'RUNX1_643_5GtoT_15_15', Gene = 'RUNX1', 
                 Target_Locus = 'RUNX1_643', Guide = 'GCATTTTCAGGAGGAAGCGA', 
                 Protospacer = 'GCATTTTCAGGAGGAAGCGA', 
                 PBS = 'CTTCCTCCTGAAAAT', Anns_RT = 'TGTCTGAAGCAATCG', 
                 RT_template = 'TGTCTGAAGCATTCG',
                 Barcode = 'TACCACACGGCATACGA',
                 Original_Target = 'GGGTGCATTTTCAGGAGGAAGCGATGGCTTCAGACAGCATATTTGAG',
                 Target_4A =       'GGGTGCATTTTCAGGAGGAAGCGAAGGCTTCAGACAGCATATTTGAG', 
                 Edited_Target =   'GGGTGCATTTTCAGGAGGAAGCGAATGCTTCAGACAGCATATTTGAG',
                 Original_Wide_Target = 'GGGTGCATTTTCAGGAGGAAGCGATGGCTTCAGACAGCATATTTGAG',
                 Wide_Target_4A = 'GGGTGCATTTTCAGGAGGAAGCGAAGGCTTCAGACAGCATATTTGAG', 
                 Edited_Wide_Target = 'GGGTGCATTTTCAGGAGGAAGCGAATGCTTCAGACAGCATATTTGAG',
                 Target_had_4A = NA, Edit_position = 5, Reference_nt = 'G', Edited_nt = 'T',
                 PBS_length = 15, RT_length = 15, GC_percent_PBS = 46.67, 
                 MLH1_Avg_Freq_TotalEdited_d7 = 0.898978816, 
                 PEmax_Avg_Freq_TotalEdited_d7 = 0.376141486,
                 MLH1_Avg_Freq_EditOnly_d7 = 0.897705471, 
                 PEmax_Avg_Freq_EditOnly_d7 = 0.376141486,
                 MLH1_Avg_Freq_TotalErrors = 0.019953115, 
                 PEmax_Avg_Freq_TotalErrors = 0.018896655, 
                 MLH1_Avg_TotalReads = 4767, PEmax_Avg_TotalReads = 5259.5,
                 mean_diff_avg_freq = 0.521563985, Target_hasATG = TRUE, Coding_hasUAA = NA,
                 Coding_hasUAG = NA, Coding_hasUGA = NA, Coding_hasStop = NA,
                 Start_of_Coding_Seq = NA, Number_to_Append = NA, Nt_to_Append = NA,
                 In_Frame_Coding_Seq = NA, 
                 In_Frame_Target_4A =     'GGGTGCATTTTCAGGAGGAAGCGAAGGCTTCAGACAGCATATTTGAG',
                 In_Frame_Edited_Target = 'GGGTGCATTTTCAGGAGGAAGCGAATGCTTCAGACAGCATATTTGAG',
                 AA_Sequence = NA, Edit_nts = 'positive_control')

peg_library_controls <- rbind(peg_library, nontarget1, nontarget2, HEK3_noedit, VEGFA, RUNX1)
```

# 6. Export data

## 6.1 Print number of edits

```{r n-edits}
# print # of edits in data
nrow(peg_library_stops)
```

## 6.2 Save a target site edit efficiency summary figure

Graph an overlay of the selected target sites' PEmax vs MLH1 KO efficiency over the original data

```{r efficiency}
# add a column to Ann's original data if the gene was used in my design
diff <- diff %>% mutate(In_New_Library = Target_Locus %in% peg_library_stops$Target_Locus)

# graph the efficiency, with genes in the new library highlighted
ggplot(diff, aes(x=PEmax_Avg_Freq_EditOnly_d7,y=MLH1_Avg_Freq_EditOnly_d7,
                    color=In_New_Library)) +
    geom_point(shape=20, aes(alpha=0.3)) +
    scale_color_manual(values=c('#595959','red')) +
    labs(title='+5 G to T Editing Efficacy by Target Site',
         x='PEmax Average
Editing Frequency',
         y='MLH1 KO Average
Editing Frequency') +
    theme_classic()
ggsave('efficiency_distribution_+5GtoT.png', plot = last_plot(), device = 'png', path = 'Summary Figures', scale = 1, width = 3.75, height = 2.5, units = "in", dpi = 300)
```

# 6.3 Export table of edit information

```{r export}
# export files
write.csv(peg_library_controls, 'MMR-responsive_+5GtoT_edit_candidates.csv', row.names=TRUE)
```